tt_eager/tensor/types.hpp

```diff
@@ -309,19 +309,127 @@ struct BorrowedStorage {
+
+    void insert_buffer_and_shape_for_device(Device* device, const OwnedBuffer buffer, const Shape shape) {
+        std::lock_guard<std::mutex> lock(mtx);
+        buffers[device->id()] = buffer;
+        shapes[device->id()] = shape;
+    }
+
+    OwnedBuffer get_buffer_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (device->id() < buffers.size()) {
+            return buffers[device->id()];;
+        }
+        return OwnedBuffer();
+    }
+
+    Shape get_tensor_shape_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (device->id() < shapes.size()) {
+            return shapes[device->id()];
+        }
+        return Shape({0xff, 0xff, 0xff, 0xff});
+    }
+
+    uint32_t num_buffers() {
+        std::lock_guard<std::mutex> lock(mtx);
+        return buffers.size();
+    }
 };
 
 struct MultiDeviceStorage {
-    std::vector<DeviceBuffer> buffers;
-    std::vector<Shape> shapes;
-
+    std::unordered_map<Device*, DeviceBuffer> buffers;
+    std::unordered_map<Device*, Shape> shapes;
+    std::mutex mtx;
     MultiDeviceStorage() = default;
+    MultiDeviceStorage(std::unordered_map<Device*, DeviceBuffer> buffers_, std::unordered_map<Device*, Shape> shapes_) : buffers(buffers_), shapes(shapes_) {}
+    MultiDeviceStorage(MultiDeviceStorage &&other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
+
+    MultiDeviceStorage(const MultiDeviceStorage &other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+    }
+
+    MultiDeviceStorage &operator=(const MultiDeviceStorage &other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
+
+    MultiDeviceStorage &operator=( MultiDeviceStorage &&other) {
+        buffers = other.buffers;
+        shapes = other.shapes;
+        return *this;
+    }
+
+    bool operator == (const MultiDeviceStorage& other) {
+        return this->buffers == other.buffers and this->shapes == other.shapes;
+    }
+
     static constexpr auto attribute_names = std::make_tuple();
     const auto attribute_values() const { return std::make_tuple(); }
+
+    void insert_buffer_and_shape_for_device(Device* device, const DeviceBuffer buffer, const Shape shape) {
+        std::lock_guard<std::mutex> lock(mtx);
+        buffers.insert({device, buffer});
+        shapes.insert({device, shape});
+    }
+
+    DeviceBuffer get_buffer_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (buffers.find(device) != buffers.end()) {
+            return buffers.at(device);
+        }
+        return DeviceBuffer();
+    }
+
+    Shape get_tensor_shape_for_device(Device* device) {
+        std::lock_guard<std::mutex> lock(mtx);
+        if (shapes.find(device) != shapes.end()) {
+            return shapes.at(device);
+        }
+        return Shape({0xff, 0xff, 0xff, 0xff});
```

Comment:
does it make more sense to throw an exception in these cases where the device isn't found?

Reply:
for both `MultiDeviceStorage` and `MultiDeviceHostStorage`

Reply:
Added to latest commits.